internal/services/mongocluster/mongo_cluster_user_resource.go

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package mongocluster
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-azure-helpers/lang/response"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/mongocluster/2025-09-01/mongoclusters"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/mongocluster/2025-09-01/users"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+type MongoClusterUserResource struct{}
+
+var _ sdk.Resource = MongoClusterUserResource{}
+
+type MongoClusterUserResourceModel struct {
+	IdentityProviderType string              `tfschema:"identity_provider_type"`
+	MongoClusterId       string              `tfschema:"mongo_cluster_id"`
+	ObjectId             string              `tfschema:"object_id"`
+	PrincipalType        string              `tfschema:"principal_type"`
+	Roles                []DatabaseRoleModel `tfschema:"role"`
+}
+
+type DatabaseRoleModel struct {
+	Database string `tfschema:"database"`
+	Role     string `tfschema:"role"`
+}
+
+func (r MongoClusterUserResource) ModelObject() interface{} {
+	return &MongoClusterUserResourceModel{}
+}
+
+func (r MongoClusterUserResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return users.ValidateUserID
+}
+
+func (r MongoClusterUserResource) ResourceType() string {
+	return "azurerm_mongo_cluster_user"
+}
+
+func (r MongoClusterUserResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		// the object_id is actually resource name
+		"object_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.IsUUID,
+		},
+
+		"mongo_cluster_id": commonschema.ResourceIDReferenceRequiredForceNew(&users.MongoClusterId{}),
+
+		"identity_provider_type": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+			ValidateFunc: validation.StringInSlice(
+				users.PossibleValuesForIdentityProviderType(),
+				false,
+			),
+		},
+
+		"principal_type": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ForceNew: true,
+			ValidateFunc: validation.StringInSlice(
+				users.PossibleValuesForEntraPrincipalType(),
+				false,
+			),
+		},
+
+		"role": {
+			Type:     pluginsdk.TypeList,
+			Required: true,
+			ForceNew: true,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*pluginsdk.Schema{
+					"database": {
+						Type:         pluginsdk.TypeString,
+						Required:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+
+					"role": {
```

Comment:
There are two "roles". It's weird. Can we give it another name?

Reply:
I think it should be fine according to [contributing doc](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/reference-naming.md).

Reply:
@teowa, this will be an issue. @neil-yechenwei is correct, we need to separate these into clear categories, a `role` block with a sub field of `role` is confusing.

Reply:
fixed, now the `role.role` is renamed to `role.name` field